packages/core/src/sdk/geolocation/useGeolocation.ts

```diff
@@ -1,42 +1,143 @@
-import { useEffect } from 'react'
+import { useEffect, useState } from 'react'
 
-import { deliveryPromise } from 'discovery.config'
-import { TIME_TO_VALIDATE_SESSION } from 'src/constants'
+import {
+  deliveryPromise,
+  session as defaultSessionValues,
+} from 'discovery.config'
 import { sessionStore, validateSession } from 'src/sdk/session'
+import { createStore, type Session } from '@faststore/sdk'
+import { useUI } from '@faststore/ui'
+import { getSavedAddress } from '../profile'
 
-async function askGeolocationConsent() {
-  const { postalCode: stalePostalCode, geoCoordinates: staleGeoCoordinates } =
-    sessionStore.read()
-
-  if (navigator?.geolocation && (!stalePostalCode || !staleGeoCoordinates)) {
-    navigator.geolocation.getCurrentPosition(
-      async ({ coords: { latitude, longitude } }) => {
-        // Revalidate the session because users can set a zip code while granting consent.
-        const revalidatedSession = sessionStore.read()
-        if (
-          revalidatedSession.postalCode ||
-          revalidatedSession.geoCoordinates
-        ) {
-          return
-        }
-
-        const newSession = {
-          ...revalidatedSession,
-          geoCoordinates: { latitude, longitude },
-        }
-        const validatedSession = await validateSession(newSession)
-        sessionStore.set(validatedSession ?? newSession)
-      }
-    )
-  }
-}
+const SECONDS = 1000
+const MINUTE = 60 * SECONDS
+
+export const geolocationStore = createStore<{
+  popupState: 'closed' | 'open'
+}>({
+  popupState: 'closed',
+})
 
 export default function useGeolocation() {
+  const { openModal, closeModal } = useUI()
+  const [sessionReady, setSessionInitialized] = useState(false)
+
   useEffect(() => {
-    if (!deliveryPromise.enabled) {
+    return sessionStore.subscribe((val) => setSessionInitialized(val.ready))
+  }, [])
+
+  useEffect(() => {
+    if (!deliveryPromise.enabled || !sessionReady) {
       return
     }
+    // async execution to use await
+    ;(async () => {
+      const sessionData = sessionStore.read()
+      const { postalCode, geoCoordinates, person } = sessionData
 
-    setTimeout(() => askGeolocationConsent(), TIME_TO_VALIDATE_SESSION)
-  }, [])
+      // always check user saved address and use it.
+      if (person?.id) {
```

Comment:
```suggestion
      if (!session.b2b && person?.id) {
```
because the b2b case is already being handled in validateSession and calling the profile query with this id will return an error 👍